core/src/main/java/jenkins/util/ScriptListener.java

```diff
@@ -0,0 +1,215 @@
+     * @param binding The script binding, or {@code null} if unavailable/inapplicable.
+     * @param feature The feature that triggered this event. Usually a fixed string or even a {@link java.lang.Class}
+     *                if that's unambiguously describing the feature (e.g., {@link hudson.cli.GroovyshCommand#getClass()}).
+     * @param context Object representing the script definition context (e.g. {@link hudson.model.Run}).
+     *                Can be {@code null} if not applicable (e.g., CLI commands not acting on jobs/builds).
+     * @param correlationId This value is used to correlate this script event to other, related script events.
+     *                      Callers are expected to provide values that allow receivers to associate script execution
+     *                      and output. Related events should have identical values.
+     * @param user If available, the user who caused this event. Can be {@code null}.
+     */
+    // TODO Should null script be allowed? Do we care about e.g. someone starting groovysh but not actually executing a command (yet)?
+    static void fireScriptExecution(@CheckForNull String script, @CheckForNull Binding binding, @NonNull Object feature, @CheckForNull Object context, @NonNull String correlationId, @CheckForNull User user) {
+        Listeners.notify(ScriptListener.class, true, listener -> listener.onScriptExecution(script, binding, feature, context, correlationId, user));
+    }
+
+    /**
+     * Fires the {@link #onScriptOutput(String, Object, Object, String, hudson.model.User)} event.
+     *
+     * @param output The output of the script.
+     * @param context Object representing the script definition context (e.g. {@link hudson.model.Run}).
+     *                Can be {@code null} if not applicable (e.g., CLI commands not acting on jobs/builds).
+     * @param correlationId This value is used to correlate this script event to other, related script events.
+     *                      Callers are expected to provide values that allow receivers to associate script execution
+     *                      and output. Related events should have identical values.
+     * @param user If available, the user for which the output was created. Can be {@code null}.
+     */
+    static void fireScriptOutput(@CheckForNull String output, @NonNull Object feature, @CheckForNull Object context, @NonNull String correlationId, @CheckForNull User user) {
+        Listeners.notify(ScriptListener.class, true, listener -> listener.onScriptOutput(output, feature, context, correlationId, user));
+    }
+
+    /**
+     * {@link java.io.Writer} that calls {@link #fireScriptOutput(String, Object, Object, String, hudson.model.User)} with the
+     * output it writes to the wrapped {@link java.io.Writer}, and otherwise just forwards {@link #flush()} and {@link #close()}.
+     */
+    class ListenerWriter extends Writer {
```

Comment:
This is implicitly `public`. If you did you not mean for this to be part of the API, use `@Restricted(NoExternalUse.class)`. Ditto other members of the interface.

Reply:
Could go either way. Realistically there are few (possibly none) possible callers of `#fire…` methods outside core (that possibly could want this), so it might be better to start off `@Restricted` and open up if necessary.